libraries/src/IoTwx.cpp

```diff
@@ -0,0 +1,368 @@
+    Change Log :
+*/
+
+#include "IoTwx.h"
+#include "nvs_flash.h"
+#include "nvs.h"
+#include "BluetoothSerial.h" // Serial Bluetooth; see tutorial on: www.circuitdigest.com 
+#include "SPIFFS.h" 
+#include <NTPClient.h>      /// https://github.com/arduino-libraries/NTPClient
+#include <WiFi.h>
+#include <WiFiUdp.h>
+#include <MQTT.h>           /// https://github.com/256dpi/arduino-mqtt
+#include <ArduinoJson.h>
+
+
+MQTTClient       mqttClient;
+WiFiUDP          ntpUDP;
+WiFiClient       networkClient;
+BluetoothSerial  btSerialConnection;      
+NTPClient        timeClient(ntpUDP);
+unsigned long    t;
+typedef uint32_t nvs_handle_t;
+unsigned long    retry_count    = 0;
+CRGB             leds[NUM_LEDS] = {LED_BLACK}; // Define the array of leds
+bool             use_led        = false;
+
+
+void init_led(){
+    // set all the LEDs to black
+    FastLED.clear();
+    FastLED.show();
+    
+    FastLED.addLeds<NEOPIXEL, NEO_PIN>(leds, NUM_LEDS);  // GRB ordering is assumed
+    use_led = true;
+}    
+
+
+void blink_led(CRGB color, int speed){
+    if (use_led) {
+        leds[0] = color;
+        
+        for (int i = 0; i < speed ; i++)
+        {
+            FastLED.setBrightness(1);
+            leds[0] = color;
+            FastLED.show();
+            delay(125);
+
+            // Now turn the LED off, then pause
+            leds[0] = LED_BLACK;
+            FastLED.show();
+            delay(250);
+        }
+    }
+}
+
+
+/**
+ * 
+ * This function waits for delay_in_sec seconds for a 
+ * BT serial connection. If there is a connection, it will
+ * accept the input as a valid JSON file and use that
+ * file for the configuration of the device.  
+ * 
+ * See === for the details of what needs to go into
+ * the config file.
+ * 
+ * The code returns true if there is proper configuration 
+ * file uploaded and successfully stored in NVS memory, 
+ * otherwise false
+ * 
+ */
+bool wait_for_bluetooth_config(const char* uuid, long last_millis, int delay_in_sec)
+{
+    File                      file;
+    StaticJsonDocument<1024>  doc;
+    char                      jsonConfig[1024]  = {'/0'};
+
+    btSerialConnection.begin(uuid); 
+    SPIFFS.begin(true);
+
+    Serial.println("[] Bluetooth Device is Ready to Pair");
+    Serial.println("[] BT initialization: 90s");
+    blink_led(LED_BT, LED_FAST);
+    
+    while (millis() - last_millis < delay_in_sec*1000) {
+        if (btSerialConnection.available()) //Check if we receive anything from Bluetooth
+        {
+            for(int i=0; btSerialConnection.available(); i++) {
+                jsonConfig[i] = char(btSerialConnection.read());
+                Serial.print(jsonConfig[i]);
+            }
+            Serial.println("[] file uploaded from BT");
+            delay(750);
+            
+            // store JSON file for future retrieval
+            Serial.println("[] trying to store file to SPIFFS");
+            file = SPIFFS.open("/config.json", FILE_WRITE);
```

Comment:
Can we make `config.json` a configurable parameter i.e. update the function signature to 

```c
bool wait_for_bluetooth_config(char* config_file, const char* uuid, long last_millis, int delay_in_sec
```

instead of hard-coding it?

Reply:
I would rather not because the following:

* basic workflow is this: 
  (1) the user turns on the device
  (2) the device is in bluetooth mode for 90s waiting for a file to be uploaded
  (3) if the user connects to the device and a file is received it is stored to the SPIFFS filesystem
IF the user does not upload a file, the system has to look for it -- therefore, if we ask the system to remember what that custom filename was, where is *that* custom name stored?  We're not using the filesystem for anything else and I don't see us using it for much of anything in the future because there is so little storage (1024K).  We *could* use it from time to time to store a few hours of measurements, but a *few* hours is all we're going to get with such a small amount to work with.  Each payload (per minute) is between 0.4K and 0.7K ... averaging that out gives us ~500 minutes or <10 hours of measurements *when using the full amount*, so I could see maybe using 1/2 or 3/4 of the SPIFFS for that purpose, but again, the user would never use it and we're not going to encourage them to do so.

All that is to say that the hardcoded name *should* be fine, given the user will never interact with it, only the system.

Thoughts on whether that makes sense or seems flawed?

Reply:
Oops my morning math was off, we'd get about 1500-2000 minutes of messages or 30-40 hours (2-3 days) until full.

Reply:
> All that is to say that the hardcoded name should be fine, given the user will never interact with it, only the system. 

👍 I now have a better understanding of why it's hard-coded. Let's leave it as is